csrc/ir/utils.cpp

```diff
@@ -1006,6 +1005,78 @@ class ValidateDomainEquivalence : private IterVisitor {
   std::unordered_set<Val*> frontier_;
 };
 
+std::vector<Statement*> next(Statement* stmt) {
+  if (stmt->isVal()) {
+    if (auto val = stmt->as<Val>()->definition()) {
+      return {val};
+    } else {
+      return {};
+    }
+  } else {
+    auto expr = stmt->as<Expr>();
+    std::vector<Statement*> inputs{
+        expr->inputs().begin(), expr->inputs().end()};
+    return inputs;
+  }
+}
+
+std::vector<Statement*> checkCycle(
+    Fusion* fusion,
+    const std::unordered_set<Statement*>& from,
+    const std::vector<Val*>& to) {
+  std::unordered_set<Statement*> path;
```

Comment:
Still trying to understand, but curious why there's only one `path` here. Wouldn't it necessary to track multiple paths?

Reply:
This is effectively a DFS approach (look at the `queue.push_front` later). The `path` tracks all the node that's on the path leading to the current node during an iteration in the while loop.

Reply:
For example, suppose an expr like `t1, t2 = op(t0)`, and if we start from `t1` and `t2` as a `to` vector, wouldn't this declare a cycle since `t0` showed up in the pats both from `t1` and `t2`?

Reply:
Pushed the above comment before realizing your response. 

I didn't realize this is a DFS traversal. Doesn't it still flag a fusion like the above case as having a cycle?

Reply:
Here's my understanding:

```
t1, t2 = op1(t0);

1. t1 and t2 are added to queue
2. t1 is visited first. Add op1 to the front of the queue. Add t1 to path. t1 stays in queue
3. op1 is visited. next_stmts is {t0}. Add t0 to queue. Add op1 to path. op1 stays in queue
4. t0 is visited. t0 is in from. t0 is dropped from queue
5. op1 is visited again. It's in path, so dropped from queue and path, and added to visited.
6. t1 is visited again. It's in path, so dropped from queue and path, and added to visited
7. t2 is visited. next_stmts is {op1}. t2 is added to path. t2 stays in queue
8. op1 is visited, but this time it's found in visited, so just dropped from queue
9. t2 is visited again. t2 is in path, so dropped from queue and added to visited
10. exists from the while loop as queue is now empty
```

Does this look correct?

And yes, you're right.  I didn't catch that a node stays in the queue when it's first visited.

Reply:
Yeah that sounds about right.